File: src/ts/examples/custom-reader.ts
/** Plugging in your own byte source by implementing RangeReader.
 *
 *      node examples/custom-reader.ts in.fcb [minX minY maxX maxY]
 *
 *  `RangeReader` is a two-method interface -- no base class to extend.
 *  Implement it and every reader, index and query works unchanged over
 *  whatever transport you have: S3, a database blob, IndexedDB, a test
 *  double. The one below wraps a Uint8Array and counts what the library
 *  asks for, which is how the request counts in the docs were measured.
 *
 *  The contract, in full (src/io/range-reader.ts):
 *
 *  * `read` resolves to EXACTLY `length` bytes at `offset`, or rejects
 *    with an `FcbError`. It never returns a short buffer and never
 *    clamps a range that runs past the end.
 *  * `size()` is SYNCHRONOUS by contract: every source learns its size
 *    when it opens, so the reader's layout arithmetic never awaits.
 */
import { readFileSync } from 'node:fs'
import { BufferedRangeReader, FcbReader, type RangeReader } from '@cityjson/flatcitybuf'

class CountingBytesReader implements RangeReader {
  readonly reads: Array<[number, number]> = []
  // Written out rather than declared as a constructor parameter
  // property: `node file.ts` strips types only, and a parameter
  // property would generate an assignment, so it is rejected.
  private readonly bytes: Uint8Array

  constructor(bytes: Uint8Array) {
    this.bytes = bytes
  }

  read(offset: number, length: number): Promise<Uint8Array> {
    this.reads.push([offset, length])
    // EXACTLY length bytes, or an error -- never a short buffer.
    if (offset + length > this.bytes.length) {
      return Promise.reject(new RangeError(`read past end: ${offset}+${length}`))
    }
    return Promise.resolve(this.bytes.subarray(offset, offset + length))
  }

  size(): number {
    return this.bytes.length
  }

  get bytesRead(): number {
    return this.reads.reduce((n, [, len]) => n + len, 0)
  }
}

const [path, ...rest] = process.argv.slice(2)
if (path === undefined || (rest.length !== 0 && rest.length !== 4)) {
  console.log('usage: node examples/custom-reader.ts <file.fcb> [minX minY maxX maxY]')
  process.exit(2)
}

const bytes = new Uint8Array(readFileSync(path))
const box =
  rest.length === 4 ? (rest.map(Number) as [number, number, number, number]) : undefined

/** Runs the query and reports what the source was actually asked for.
 *  `wrap` decides whether the raw counter is handed to the reader
 *  directly or through a BufferedRangeReader. */
async function run(label: string, wrap: boolean): Promise<void> {
  const source = new CountingBytesReader(bytes)
  const reader = await FcbReader.fromReader(
    wrap ? new BufferedRangeReader(source) : source,
  )
  try {
    const info = reader.header.info
    if (box === undefined) {
      console.log(
        `opened: ${info.featuresCount} features, ${source.size()} bytes, ` +
          `${source.reads.length} read(s)`,
      )
      console.log('pass a bbox to see how little a query actually reads')
      return
    }

    const before: [number, number] = [source.reads.length, source.bytesRead]
    let n = 0
    for await (const feature of await reader.select({ spatial: { kind: 'bbox', value: box } })) {
      if (n < 5 && !wrap) console.log(`  ${feature.id}`)
      n += 1
    }
    if (n > 5 && !wrap) console.log(`  ... ${n - 5} more`)

    const nReads = source.reads.length - before[0]
    const nBytes = source.bytesRead - before[1]
    const pct = ((100 * nBytes) / source.size()).toFixed(1)
    console.log(
      `${label}: ${n} hit(s), ${nReads} read(s), ${nBytes} bytes (${pct}% of the file)`,
    )
  } finally {
    await reader.close()
  }
}

// `fromReader` uses the source EXACTLY as given -- it inserts no
// buffering decorator, deliberately, so a request count stays honest and
// tunable. A sequential read costs TWO reads per feature (a 4-byte size
// prefix, then the body), which is why the raw number below is roughly
// 2x the hit count. Callers over a chatty transport -- HTTP above all --
// wrap the source themselves, exactly as `fromUrl` does internally.
await run('raw     ', false)
await run('buffered', true)

File: src/py/examples/custom_reader.py
"""Plugging in your own byte source by implementing RangeReader.

    python examples/custom_reader.py in.fcb [minx miny maxx maxy]

`RangeReader` is a Protocol -- two methods, no base class to inherit.
Implement it and every reader, index and query works unchanged over
whatever transport you have: S3, a database blob, an mmap, a test
double. The one below wraps a file and counts what the library asks
for, which is how the request counts in the docs were measured.

The contract, in full (range_reader.py):

* `read(offset, length)` returns EXACTLY `length` bytes, unless the
  range crosses the end of the resource, in which case it returns
  exactly the bytes that exist (possibly zero).
* `length == 0` returns `b""` WITHOUT contacting the transport.
* Errors raise `FcbError`. Returning garbage is not an option --
  callers cannot tell it from data.
* Instances are NOT thread-safe.
"""

from __future__ import annotations

import sys

import flatcitybuf as fcb


class CountingFileReader:
    """A RangeReader over a local file that records every read."""

    def __init__(self, path: str) -> None:
        self._fh = open(path, "rb")
        self._fh.seek(0, 2)
        self._size = self._fh.tell()
        self.reads: list[tuple[int, int]] = []

    def read(self, offset: int, length: int) -> bytes:
        if length == 0:
            return b""  # the contract: no transport call at all
        self.reads.append((offset, length))
        self._fh.seek(offset)
        return self._fh.read(length)

    def total_size(self) -> int:
        return self._size

    def close(self) -> None:
        self._fh.close()

    @property
    def bytes_read(self) -> int:
        return sum(n for _, n in self.reads)


def main(argv: list[str]) -> int:
    path = argv[0]
    box = tuple(float(v) for v in argv[1:5]) if len(argv) == 5 else None

    source = CountingFileReader(path)
    try:
        reader = fcb.FcbReader.open(source)
        info = reader.header.info
        print(
            f"opened: {info.features_count} features, "
            f"{source.total_size()} bytes on disk, "
            f"{len(source.reads)} read(s) so far"
        )

        if box is None:
            print("pass a bbox to see how little a query actually reads")
            return 0

        before = len(source.reads), source.bytes_read
        hits = fcb.search_rtree(
            reader.range_reader,
            reader.header.layout.rtree_begin,
            info.features_count,
            info.index_node_size,
            box,  # type: ignore[arg-type]
        )
        for hit in hits[:5]:
            print(f"  {reader.feature_at(hit).id}")
        if len(hits) > 5:
            print(f"  ... {len(hits) - 5} more")

        n = len(source.reads) - before[0]
        b = source.bytes_read - before[1]
        pct = 100.0 * b / source.total_size()
        print(
            f"{len(hits)} hit(s): {n} read(s), {b} bytes "
            f"({pct:.1f}% of the file)"
        )
        return 0
    finally:
        source.close()


if __name__ == "__main__":
    if len(sys.argv) not in (2, 6):
        print(__doc__)
        raise SystemExit(2)
    raise SystemExit(main(sys.argv[1:]))
